### PR TITLE
Generate JSON Schema for both Resolved Telemetry Schema and Resolved Registry

### PR DESCRIPTION
**Pull request by lquerel**
This PR uses [schemars](https://github.com/GREsau/schemars) to automatically generate JSON schema from Serde-annotated Rust structs. 

**This PR will serve as a basis for future discussions and the generation of documentation on the topic of resolved registry and schema.**

Two types of resolved assets coexist:
- **Resolved Registry**: A resolved view of a semantic convention registry. This is basically a resolved/flattened version of the semantic convention files. The command `weaver registry resolve` can be used to generate this type of resolved asset. The command `weaver registry json-schema` generates the JSON Schema representation of a `ResolvedRegistry` data structure.
- **Resolved Telemetry Schema**: An extension of the existing Telemetry Schema containing a catalog of attributes, registries, etc., in addition to the original `versions` section. This type of schema is still in discussion.

This PR also improves the output of commands such as `weaver registry resolve`, which can now be directly piped to a `jq` command without having to use the `--quiet` parameter (see #189).

Note 1: `TemplateRegistry` and `TemplateGroup` have been renamed to `ResolvedRegistry` and `ResolvedGroup` because it's a more logical name. This renaming has no impact on the templates and the policies. I might move these definitions into a dedicated crate.

Note 2: A conversation has been initiated to talk about resolved registry and resolved telemetry schema, see https://cloud-native.slack.com/archives/C0697EXNTL3/p1717183349294419?thread_ts=1717104224.284179&cid=C0697EXNTL3

### CODE REVIEW

File: crates/weaver_common/src/lib.rs [hunk 1]
```diff
@@ -47,6 +47,9 @@ pub trait Logger {
 
     /// Logs a message without icon.
     fn log(&self, message: &str);
+
+    /// Mute all the messages except for the warnings and errors.
+    fn mute(&self);
```
Comment (lquerel) [inline crates/weaver_common/src/lib.rs:52]: Introduced this new function to fix #184 and to simplify testing.

File: src/diagnostic/init.rs [hunk 2]
```diff
@@ -3,7 +3,7 @@
 //! Initializes a `diagnostic_templates` directory to define or override diagnostic output formats.
 
 use crate::diagnostic::{Error, DEFAULT_DIAGNOSTIC_TEMPLATES};
-use crate::DiagnosticArgs;
+use crate::{DiagnosticArgs, ExitDirectives};
```
Comment (lquerel) [inline src/diagnostic/init.rs:6]: ExitDirectives is a way for commands to inform the main program on how it should behave once the command has been executed (e.g., suggested exit code, quiet mode).

File: src/format.rs [hunk 3]
```diff
@@ -0,0 +1,25 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! Supported output formats for the resolved registry and telemetry schema
+
+use clap::ValueEnum;
+use serde::Serialize;
+
+/// Supported output formats for the resolved schema
+#[derive(Debug, Clone, ValueEnum)]
+pub(crate) enum Format {
```
Comment (lquerel) [inline src/format.rs:10]: Moved `Format` in a more global location as this structure is now used from multiple sub-commands.

File: src/registry/resolve.rs [hunk 4]
```diff
@@ -78,7 +65,10 @@ pub(crate) fn command(
     logger: impl Logger + Sync + Clone,
     cache: &Cache,
     args: &RegistryResolveArgs,
-) -> Result<(), DiagnosticMessages> {
+) -> Result<ExitDirectives, DiagnosticMessages> {
```
Comment (lquerel) [inline src/registry/resolve.rs:68]: Now this command can only display the resolved registry. See `weaver schema resolve` to display the resolved telemetry schema (this part is still experimental).

File: crates/weaver_common/src/lib.rs [hunk 5]
```diff
@@ -55,6 +58,7 @@ pub trait Logger {
 pub struct ConsoleLogger {
     logger: Arc<Mutex<paris::Logger<'static>>>,
     debug_level: u8,
+    mute: Arc<Mutex<bool>>,
```
Comment (jsuereth) [inline crates/weaver_common/src/lib.rs:61]: I'd just use an atomic here.

File: crates/weaver_common/src/lib.rs [hunk 6]
```diff
@@ -64,14 +68,16 @@ impl ConsoleLogger {
...
+            mute: Arc::new(Mutex::new(false)),
         }
     }
 }
 
 impl Logger for ConsoleLogger {
     /// Logs an trace message (only with debug enabled).
     fn trace(&self, message: &str) {
-        if self.debug_level > 0 {
+        let mute = *self.mute.lock().expect("Failed to lock mute");
```
Comment (jsuereth) [inline crates/weaver_common/src/lib.rs:79]: Why do we need to lock? Just check current value.

Also if you set this early, you may even be able to lazy-read in practice...

File: src/schema/mod.rs [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+pub enum SchemaSubCommand {
+    /// Resolves a telemetry schema.
+    ///
+    /// Rego policies present in the registry or specified using -p or --policy will be automatically validated by the policy engine before the artifact generation phase.
+    ///
+    /// Note: The `-d` and `--registry-git-sub-dir` options are only used when the registry is a Git URL otherwise these options are ignored.
+    ///
+    /// The process exits with a code of 0 if the resolution is successful.
+    #[clap(verbatim_doc_comment)]
+    Resolve(SchemaResolveArgs),
```
Comment (jsuereth) [inline src/schema/mod.rs:35]: I'm not sure I understand what this is meant to do.

**Review by jsuereth — COMMENTED** [review]
Are we declaring the difference between schema + registry the difference between using optimisation on attributes vs. not?

At this point, I'm still a bit confused on terminology and I think maybe we should discuss that first.

Also have a minor nit about threading choices.

File: crates/weaver_common/src/lib.rs (hunk 5 above)
Comment (lquerel) [reply crates/weaver_common/src/lib.rs:61]: I agree, a mutex is overkill in this context. An atomic is indeed enough, I don't know why I didn't use it in the first place.

File: src/schema/mod.rs (hunk 7 above)
Comment (lquerel) [reply src/schema/mod.rs:35]: We need to discuss that in a more interactive way.

**Comment by lquerel** [comment]
> At this point, I'm still a bit confused on terminology and I think maybe we should discuss that first.

Yes, I think the best way to discuss this is through chat or via Zoom.

File: crates/weaver_common/src/lib.rs (hunk 5 above)
Comment (lquerel) [reply crates/weaver_common/src/lib.rs]: Replaced the mutex by an atomic in [d14e711](https://github.com/open-telemetry/weaver/pull/187/commits/d14e711c0028d10c21b260cbf287db549b6dc8d4)

File: crates/weaver_common/src/lib.rs (hunk 6 above)
Comment (lquerel) [reply crates/weaver_common/src/lib.rs]: Replaced the mutex by an atomic in [d14e711](https://github.com/open-telemetry/weaver/pull/187/commits/d14e711c0028d10c21b260cbf287db549b6dc8d4)

File: src/schema/mod.rs (hunk 7 above)
Comment (lquerel) [reply src/schema/mod.rs]: Conversation initiated here https://cloud-native.slack.com/archives/C0697EXNTL3/p1717183349294419?thread_ts=1717104224.284179&cid=C0697EXNTL3

**Review by jsuereth — APPROVED** [review]

**Pull request merged**